### PR TITLE
Add `_.zipWith`.

### PR DESCRIPTION
**Pull request by thirdcreed**
First iteration of `_.zipWith`.

### CODE REVIEW

File: lodash.js [hunk 1]
```diff
@@ -5649,6 +5649,7 @@
      * _.zipObject(['fred', 'barney'], [30, 40]);
      * // => { 'fred': 30, 'barney': 40 }
      */
+
```
Comment (jdalton) [inline lodash.js]: whitespace nit.

File: lodash.js [hunk 2]
```diff
@@ -5667,6 +5668,15 @@
       }
       return result;
     }
+      
+
```
Comment (jdalton) [inline lodash.js]: Can you add a comment block here.

File: lodash.js [hunk 3]
```diff
@@ -5667,6 +5668,15 @@
       }
       return result;
     }
+      
+    
+    var zipWith = restParam(
+      function zipwith(foldFunction, lists) {
```
Comment (jdalton) [inline lodash.js]: no named function expression for now (it's not consistent with our style).

File: lodash.js [hunk 4]
```diff
@@ -5667,6 +5668,15 @@
...
+        var zippedLists = _.zip.apply(this,lists)
```
Comment (jdalton) [inline lodash.js]: `unzip` will except an array. also use `unzip(lists)` and not `_.unzip(lists)`.

File: lodash.js [hunk 5]
```diff
@@ -5667,6 +5668,15 @@
...
+          return _.map(zippedLists, function reduceZips(list) {
```
Comment (jdalton) [inline lodash.js]: named expression.

**Comment by jdalton** [comment]
You only need to modify the lodash.src.js file and add tests (no need to modify the lodash.js as that's built from lodash.src.js).

File: test/test.js [hunk 6]
```diff
@@ -16630,6 +16630,20 @@
       strictEqual(_.object, _.zipObject);
     });
   }());
+    
+  /*--------------------------------------------------------------------------*/
+  QUnit.module('lodash.zipWith');
+  
+    (function() {
```
Comment (jdalton) [inline test/test.js]: Spacing mixup, it's two space (not tabs).

File: lodash.js (hunk 1 above)
Comment (thirdcreed) [inline lodash.js]: Fixed in lodash.src.js

File: lodash.js (hunk 2 above)
Comment (thirdcreed) [inline lodash.js]: Fixed in lodash.src.js

File: lodash.js (hunk 3 above)
Comment (thirdcreed) [inline lodash.js]: Fixed in lodash.src.js

File: lodash.js (hunk 4 above)
Comment (thirdcreed) [inline lodash.js]: Fixed in lodash.src.js

File: lodash.js (hunk 5 above)
Comment (thirdcreed) [inline lodash.js]: Fixed in lodash.src.js

**Comment by thirdcreed** [comment]
All right, let me know if I should iterate more, i'm uncertain of the expected behavior of zipWith when the lists have unequal lengths,  right now, i'm letting the reduce function decide for me

Haskell only evaluates up to the length of the shortest list.  

I'm don't have a strong opinion as long as its behavior is predictable.

File: lodash.src.js [hunk 7]
```diff
@@ -6022,7 +6022,7 @@
      * _.zipObject(['fred', 'barney'], [30, 40]);
      * // => { 'fred': 30, 'barney': 40 }
      */
-    function zipObject(props, values) {
+     function zipObject(props, values) {
```
Comment (jdalton) [inline lodash.src.js]: No need to change the space on `zipObject`.

File: lodash.src.js [hunk 8]
```diff
@@ -6042,6 +6042,14 @@
     }
 
     /*------------------------------------------------------------------------*/
+    var zipWith = restParam(
+        function(foldFunc, lists) {
```
Comment (jdalton) [inline lodash.src.js]: 2 space indents please.

File: lodash.src.js [hunk 9]
```diff
@@ -11975,6 +11983,8 @@
     lodash.xor = xor;
     lodash.zip = zip;
     lodash.zipObject = zipObject;
+    lodash.zipWith = zipWith;
+
```
Comment (jdalton) [inline lodash.src.js]: No need for the extra newline.

File: test/test.js [hunk 10]
```diff
@@ -16630,6 +16630,19 @@
...
+  QUnit.module('lodash.zipWith');
+  
+  (function() {
+    test('should combine values in lists with given function ',2, function(){
+      var list1 = [1,2,3];
+      var list2 = [1,2,3];
+
+      deepEqual(_.zipWith(_.add,list1,list2),[2,4,6]);
+      deepEqual(_.zipWith(_.add,list1,[]),[1,2,3]);  
+      });
```
Comment (jdalton) [inline test/test.js]: indent nitpick.

File: lodash.src.js [hunk 11]
```diff
@@ -6042,6 +6042,14 @@
     }
 
     /*------------------------------------------------------------------------*/
+    var zipWith = restParam(
+        function(foldFunc, lists) {
+          return _.map(unzip(lists), function(list) {
```
Comment (d10) [inline lodash.src.js]: We can probably use simpler versions like `arrayMap` and friends instead of `_.map`.

**Comment by thirdcreed** [comment]
Eesh. Thanks for you patience on these.  I'll replace with arrayMap now.

File: lodash.src.js [hunk 12]
```diff
@@ -6042,6 +6042,14 @@
     }
 
     /*------------------------------------------------------------------------*/
+    var zipWith = restParam(
```
Comment (jdalton) [inline lodash.src.js]: Can you add a doc block with a usage example.

**Comment by jdalton** [comment]
@thirdcreed Awesome thank you! This won't make it in the 3.7.0 release (due out in a day or so) but will be added after.

**Comment by jdalton** [comment]
Have you signed [our CLA](http://dojofoundation.org/about/claCheck)?

**Comment by thirdcreed** [comment]
Of course! My pleasure. I just signed the CLA.

**Comment by jacobp100** [comment]
How does this differ from _.merge other than perf?

**Comment by jdalton** [comment]
@jacobp100 It's not like `_.merge`. See the doc description:
> Combines elements of given arrays, like `_.zip` but with a function   specifying how they should be combined.

**Comment by thirdcreed** [comment]
_.merge takes two objects as input, looks through each for a key name, then merges arrays found in the values of matching keys,  it can use a function to customize the merging process, that function's inputs are the values in the object keys.

_.zipWith takes a function and n lists as input .  It outputs  a list of each of those with identical indices after transforming with a reduce function.

Incidentally , Since arrays are objects whose indices are keys, merge can be used with similar functionality.  It's a lot of overhead to go through all the other merge logic to zip lists, but if it's judged to be worth the trouble that's fine, it just gives people who come from other functional languages an abstraction they are used to, and it explicitly rather than incidentally merges lists.

```js
_.merge([1,2,3],[3,4,5],[1,2,3],[10,20,30],function(a,b){
 return a - b;
})

 ==

_.zipWith(function(a,b){
 return a - b;
},[1,2,3],[3,4,5],[1,2,3],[10,20,30],)

```

**Comment by thirdcreed** [comment]
I would actually like to have two separate functions for this, so that they can develop independently, zipWith has one well understood functionality, and is unlikely to change much. merge already has deep props slated to be implemented on the roadmap.

**Comment by jdalton** [comment]
Ya, they may happen to be able to produce the same results but they are doing totally different things. The roadmap note for `_.merge` is about it operating on functions.

File: lodash.src.js [hunk 13]
```diff
@@ -6042,6 +6042,30 @@
...
+     * @static
+     * @memberOf _
+     * @category Array
+     * @param {Function} accumulator Function used to reduce zipped elements
+     * @param {...Array} [arrays] Arrays to be zipped with accumulator
+     * @returns {Array} Returns new array of accumulated groups
+     * @example
+     *
+     * _.zipWith(_.add,[1,2,3],[10,20,30],[100,200,300]) 
+     * // => [111,222,333]
```
Comment (jdalton) [inline lodash.src.js]: Can you flip the arg order on this. In lodash-fp we'll have it be iteratee first but to align with lodash primary it should be at the end.

Comment (thirdcreed) [inline lodash.src.js]: Sure, will do.

**Comment by jdalton** [comment]
@thirdcreed Holy crap you're from Norman?! I went to OU!

**Comment by thirdcreed** [comment]
That's awesome! I worked on the research campus at OU for a year or so. I actually grew up in a tiny town in western Oklahoma, but I've been in Norman for the past couple years.

**Comment by jdalton** [comment]
I grew up in Lawton! Okies! Woop woop!

**Comment by thirdcreed** [comment]
So I  actually lived in Lawton (by the pizza hut on cache road) for three years while I taught music in an intermediate school in Cache  Hey if you're back in OK sometime you should drop by our usergroup: http://okcjs.com/  You have an open invitation.

**Comment by thirdcreed** [comment]
The group leader Jesse Harlin(@the_simian) says he'll cover travel expenses.  We'd love to have you.

**Pull request merged**